### PR TITLE
fix(cache): Store cache with sorted object keys

### LINKED ISSUE #17603: `prCache` within repo cache is written while being out of date
### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us what version of Renovate you run.

latest

### If you're self-hosting Renovate, select which platform you are using.

github.com

### If you're self-hosting Renovate, tell us what version of the platform you run.

_No response_

### Was this something which used to work for you, and then stopped?

It used to work, and then stopped

### Describe the bug

`prCache` within repo cache is written while being out of date which causes unnecessary sequential repo cache write. 

Steps to reproduce -
1. Have a valid repo cache
2. Change the repo in a way that will cause Renovate to open/update a PR
3. Run renovate
3.1. Repo cache is being written (with out of date `prCache`)

4. Run renovate again
  4.1. Renovate does not preform any updates
  4.2. Repo cache is being written again (with actual up to date `prCache`)

For example, 
Renovate force pushes a PR -

image

Out of date saved `prCache` -

```
"prCache": {
                 "9": {
                      "head": {
                                     "sha": "25ba43e3ed36305df07065cb5d7af4906248bdd8",
                       ....


```

Minimal reproductions repo -
Any will do, I have used the following
https://github.com/ladzaretti/alpine-node

### Relevant debug logs

Note that the logged `saving repo cache` here was added locally for the purpose of this issue.

<details><summary>Logs</summary>

```
DEBUG: Performing blobless clone (repository=ladzaretti/alpine-node)
DEBUG: git clone completed (repository=ladzaretti/alpine-node)
       "durationMs": 1828
DEBUG: latest repository commit (repository=ladzaretti/alpine-node)
       "latestCommit": {
         "hash": "c7675591cf1e68ca9e87a00b8ff8da4245f5e259",
         "date": "2022-09-03T22:22:26+03:00",
         "message": "Update Dockerfile",
         "refs": "HEAD -> main, origin/main, origin/HEAD",
         "body": "",
         "author_name": "ladzaretti",
         "author_email": "g.ladzaretti@gmail.com"
       }
DEBUG: latest commit (repository=ladzaretti/alpine-node)
       "branchName": "main",
       "latestCommitDate": "2022-09-03T22:22:26+03:00"
...
DEBUG: git commit (repository=ladzaretti/alpine-node, branch=renovate/ubuntu-22.x)
       "deletedFiles": [],
       "ignoredFiles": [],
       "result": {
         "author": null,
         "branch": "renovate/ubuntu-22.x",
         "commit": "ac86adffbb59771199d21bf5550fbed5e75fffb6",
         "root": false,
         "summary": {"changes": 1, "insertions": 1, "deletions": 1}
       }
DEBUG: Pushing branch renovate/ubuntu-22.x (repository=ladzaretti/alpine-node, branch=renovate/ubuntu-22.x)
DEBUG: git push (repository=ladzaretti/alpine-node, branch=renovate/ubuntu-22.x)
       "result": {
         "pushed": [],
         "ref": {"local": "refs/remotes/origin/renovate/ubuntu-22.x"},
         "remoteMessages": {"all": []}
       }
 INFO: Branch updated (repository=ladzaretti/alpine-node, branch=renovate/ubuntu-22.x)
       "commitSha": "ac86adffbb59771199d21bf5550fbed5e75fffb6"
DEBUG: Checking if we can automerge branch (repository=ladzaretti/alpine-node, branch=renovate/ubuntu-22.x)
DEBUG: mergeStatus=no automerge (repository=ladzaretti/alpine-node, branch=renovate/ubuntu-22.x)
DEBUG: Ensuring PR (repository=ladzaretti/alpine-node, branch=renovate/ubuntu-22.x)
DEBUG: There are 0 errors and 0 warnings (repository=ladzaretti/alpine-node, branch=renovate/ubuntu-22.x)
DEBUG: getBranchPr(renovate/ubuntu-22.x) (repository=ladzaretti/alpine-node, branch=renovate/ubuntu-22.x)
DEBUG: findPr(renovate/ubuntu-22.x, undefined, open) (repository=ladzaretti/alpine-node, branch=renovate/ubuntu-22.x)
DEBUG: Found PR #9 (repository=ladzaretti/alpine-node, branch=renovate/ubuntu-22.x)
DEBUG: Found existing PR (repository=ladzaretti/alpine-node, branch=renovate/ubuntu-22.x)
DEBUG: Processing existing PR (repository=ladzaretti/alpine-node, branch=renovate/ubuntu-22.x)
DEBUG: PR body changed (repository=ladzaretti/alpine-node, branch=renovate/ubuntu-22.x)
       "prTitle": "Update ubuntu Docker tag to v22"
DEBUG: updatePr(9, Update ubuntu Docker tag to v22, body) (repository=ladzaretti/alpine-node, branch=renovate/ubuntu-22.x)
DEBUG: PR updated (repository=ladzaretti/alpine-node, branch=renovate/ubuntu-22.x)
       "pr": 9
 INFO: PR updated (repository=ladzaretti/alpine-node, branch=renovate/ubuntu-22.x)
       "pr": 9,
       "prTitle": "Update ubuntu Docker tag to v22"
DEBUG: PR is not configured for automerge (repository=ladzaretti/alpine-node, branch=renovate/ubuntu-22.x)
DEBUG: getBranchPr(renovate/ubuntu-22.x) (repository=ladzaretti/alpine-node)
DEBUG: findPr(renovate/ubuntu-22.x, undefined, open) (repository=ladzaretti/alpine-node)
DEBUG: Found PR #9 (repository=ladzaretti/alpine-node)
DEBUG: Ensuring Dependency Dashboard (repository=ladzaretti/alpine-node)
DEBUG: ensureIssue(Dependency Dashboard) (repository=ladzaretti/alpine-node)
DEBUG: Patching issue (repository=ladzaretti/alpine-node)
DEBUG: Issue updated (repository=ladzaretti/alpine-node)
DEBUG: saving repo cache (repository=ladzaretti/alpine-node)
       "cache": {
         ...
                 "9": {
                   "id": 1045540679,
                   "node_id": "PR_kwDOHzClJ84-Ua9H",
                   "number": 9,
                   "state": "open",
                   "locked": false,
                   "title": "Update ubuntu Docker tag to v22",
                   "user": {
                     "login": "ladzaretti",
                     "id": 24865016,
                     "node_id": "MDQ6VXNlcjI0ODY1MDE2",
                     "gravatar_id": "",
                     "type": "User",
                     "site_admin": false
                   },
                   "created_at": "2022-09-03T18:52:20Z",
                   "updated_at": "2022-09-03T19:14:23Z",
                   "closed_at": null,
                   "merged_at": null,
                   "merge_commit_sha": "0e87ee290ea6c9240ba032b9f695897e97cb13c0",
                   "assignee": null,
                   "assignees": [],
                   "requested_reviewers": [],
                   "requested_teams": [],
                   "labels": [],
                   "milestone": null,
                   "draft": false,
                   "head": {
                     "label": "ladzaretti:renovate/ubuntu-22.x",
                     "ref": "renovate/ubuntu-22.x",
                     "sha": "25ba43e3ed36305df07065cb5d7af4906248bdd8",
                     "user": {
                       "login": "ladzaretti",
                       "id": 24865016,
                       "node_id": "MDQ6VXNlcjI0ODY1MDE2",
                       "gravatar_id": "",
                       "type": "User",
                       "site_admin": false
                     },
                     "repo": {
                       "id": 523281703,
                       "node_id": "R_kgDOHzClJw",
                       "name": "alpine-node",
                       "full_name": "ladzaretti/alpine-node",
                       "private": false,
                       "owner": {
                         "login": "ladzaretti",
                         "id": 24865016,
                         "node_id": "MDQ6VXNlcjI0ODY1MDE2",
                         "gravatar_id": "",
                         "type": "User",
                         "site_admin": false
                       },
                       "description": null,
                       "fork": false,
                       "created_at": "2022-08-10T09:32:23Z",
                       "updated_at": "2022-08-10T09:34:43Z",
                       "homepage": null,
                       "size": 6,
                       "stargazers_count": 0,
                       "watchers_count": 0,
                       "language": "Dockerfile",
                       "has_issues": true,
                       "has_projects": true,
                       "has_downloads": true,
                       "has_wiki": true,
                       "has_pages": false,
                       "forks_count": 1,
                       "mirror_url": null,
                       "archived": false,
                       "disabled": false,
                       "open_issues_count": 2,
                       "license": null,
                       "allow_forking": true,
                       "is_template": false,
                       "web_commit_signoff_required": false,
                       "topics": [],
                       "visibility": "public",
                       "forks": 1,
                       "open_issues": 2,
                       "watchers": 0,
                       "default_branch": "main"
                     }
                   },
                   "base": {
                     "label": "ladzaretti:main",
                     "ref": "main",
                     "sha": "55cae51128259cfb8047bff056898b7b20d620b7",
                     "user": {
                       "login": "ladzaretti",
                       "id": 24865016,
                       "node_id": "MDQ6VXNlcjI0ODY1MDE2",
                       "gravatar_id": "",
                       "type": "User",
                       "site_admin": false
                     },
                     "repo": {
                       "id": 523281703,
                       "node_id": "R_kgDOHzClJw",
                       "name": "alpine-node",
                       "full_name": "ladzaretti/alpine-node",
                       "private": false,
                       "owner": {
                         "login": "ladzaretti",
                         "id": 24865016,
                         "node_id": "MDQ6VXNlcjI0ODY1MDE2",
                         "gravatar_id": "",
                         "type": "User",
                         "site_admin": false
                       },
                       "description": null,
                       "fork": false,
                       "created_at": "2022-08-10T09:32:23Z",
                       "updated_at": "2022-08-10T09:34:43Z",
                       "homepage": null,
                       "size": 6,
                       "stargazers_count": 0,
                       "watchers_count": 0,
                       "language": "Dockerfile",
                       "has_issues": true,
                       "has_projects": true,
                       "has_downloads": true,
                       "has_wiki": true,
                       "has_pages": false,
                       "forks_count": 1,
                       "mirror_url": null,
                       "archived": false,
                       "disabled": false,
                       "open_issues_count": 2,
                       "license": null,
                       "allow_forking": true,
                       "is_template": false,
                       "web_commit_signoff_required": false,
                       "topics": [],
                       "visibility": "public",
                       "forks": 1,
                       "open_issues": 2,
                       "watchers": 0,
                       "default_branch": "main"
                     }
                   },
                   "author_association": "OWNER",
                   "auto_merge": null,
                   "active_lock_reason": null,
                   "bodyStruct": {
                     "hash": "72fbc5f1fc359bd8c9ba4122476bf810bb114619a41cfc83fd7563c32c8c9321",
                     "debugData": {
                       "createdInVer": "0.0.0-semantic-release",
                       "updatedInVer": "0.0.0-semantic-release"
                     }
                   }
                 }
               },
               "lastModified": "2022-09-03T19:14:23Z"
             }
           }
         },
         "branches": [
           {
             "branchName": "renovate/ubuntu-22.x",
             "baseBranch": "main",
             "sha": "ac86adffbb59771199d21bf5550fbed5e75fffb6",
             "parentSha": "c7675591cf1e68ca9e87a00b8ff8da4245f5e259",
             "prNo": 9,
             "automerge": false,
             "isModified": false,
             "upgrades": [
               {
                 "datasource": "docker",
                 "depName": "ubuntu",
                 "fixedVersion": "16.04",
                 "currentVersion": "16.04",
                 "newVersion": "jammy"
               }
             ],
             "branchFingerprint": "249baaf39234c2f14b7c2e16e37a5ed3bb1d0b4f9b9d43af7057a6efadaadedbcea5fb03155344b481f6fa00187272eea91c585ff680f53a04fa6901e6a70bb0"
           }
         ],
         "gitConflicts": {
           "main": {
             "targetBranchSha": "c7675591cf1e68ca9e87a00b8ff8da4245f5e259",
             "sourceBranches": {
               "renovate/ubuntu-22.x": {
                 "sourceBranchSha": "ac86adffbb59771199d21bf5550fbed5e75fffb6",
                 "isConflicted": false
               }
             }
           }
         }
       }
DEBUG: Removing any stale branches (repository=ladzaretti/alpine-node)
DEBUG: config.repoIsOnboarded=true (repository=ladzaretti/alpine-node)
DEBUG: Branch lists (repository=ladzaretti/alpine-node)
       "branchList": ["renovate/ubuntu-22.x"],
       "renovateBranches": ["renovate/ubuntu-22.x"]
DEBUG: remainingBranches= (repository=ladzaretti/alpine-node)
DEBUG: No branches to clean up (repository=ladzaretti/alpine-node)
DEBUG: PackageFiles.clear() - Package files deleted (repository=ladzaretti/alpine-node)
       "baseBranches": ["main"]
DEBUG: Branch summary (repository=ladzaretti/alpine-node)
       "cacheModified": true,
       "baseBranches": [{"branchName": "main", "sha": "c7675591cf1e68ca9e87a00b8ff8da4245f5e259"}],
       "branches": [
         {
           "branchName": "renovate/ubuntu-22.x",
           "branchSha": "ac86adffbb59771199d21bf5550fbed5e75fffb6",
           "baseBranch": "main",
           "baseBranchSha": "c7675591cf1e68ca9e87a00b8ff8da4245f5e259",
           "automerge": false,
           "isModified": false
         }
       ],
       "inactiveBranches": []
DEBUG: Renovate repository PR statistics (repository=ladzaretti/alpine-node)
       "stats": {"total": 8, "open": 1, "closed": 7, "merged": 0}
DEBUG: Repository result: done, status: onboarded, enabled: true, onboarded: true (repository=ladzaretti/alpine-node)
DEBUG: Repository timing splits (milliseconds) (repository=ladzaretti/alpine-node)
       "splits": {"init": 1895, "extract": 3432, "lookup": 46, "onboarding": 0, "update": 5822},
       "total": 12658
DEBUG: http statistics (repository=ladzaretti/alpine-node)
       "urls": {
         "https://api.github.com/graphql (POST,200)": 2,
         "https://api.github.com/repos/ladzaretti/alpine-node/branches/main/protection (GET,404)": 1,
         "https://api.github.com/repos/ladzaretti/alpine-node/contents/renovate.json (GET,200)": 1,
         "https://api.github.com/repos/ladzaretti/alpine-node/issues/3 (GET,200)": 2,
         "https://api.github.com/repos/ladzaretti/alpine-node/issues/3 (PATCH,200)": 1,
         "https://api.github.com/repos/ladzaretti/alpine-node/pulls (GET,200)": 1,
         "https://api.github.com/repos/ladzaretti/alpine-node/pulls/9 (PATCH,200)": 1
       },
       "hostStats": {"api.github.com": {"requestCount": 9, "requestAvgMs": 485, "queueAvgMs": 0}},
       "totalRequests": 9
 INFO: Repository finished (repository=ladzaretti/alpine-node)
       "durationMs": 12658
DEBUG: dns cache (repository=ladzaretti/alpine-node)
       "hosts": []
DEBUG: Renovate exiting
Done in 26.58s.

Process finished with exit code 0


```

</details>

### Have you created a minimal reproduction repository?

I have linked to a minimal reproduction repository in the bug description

### PR DESCRIPTION
**Pull request by zharinov**
## Changes

Deep sort objects keys before `JSON.stringify`. This should help to avoid unnecessary disk writes of the essentially same data. Arrays are left unsorted as it may be unsafe.

## Context

- Fixes: #17603

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [x] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

**Comment by zharinov** [comment]
I've chosen `sortobject` over `json-stable-sort` since it's zero-dependency library and it was updated recently. Please object if you have a better choice.

File: lib/util/cache/repository/impl/local.spec.ts [hunk 1]
```diff
@@ -306,7 +307,19 @@ describe('util/cache/repository/impl/local', () => {
...
+    expect(localRepoCache.isModified()).toBeFalse();
+
+    const data = localRepoCache.getData();
+
+    // Additionaly test for object keys order-independency
+    delete data.configFileName;
+    delete data.semanticCommits;
+    data.semanticCommits = 'enabled';
+    data.configFileName = 'renovate.json';
     expect(localRepoCache.isModified()).toBeFalse();
```
Comment (zharinov) [inline lib/util/cache/repository/impl/local.spec.ts:323]: It fails with the current implementation

File: lib/util/cache/repository/impl/local.spec.ts [hunk 2]
```diff
@@ -313,4 +313,31 @@ describe('util/cache/repository/impl/local', () => {
...
+    delete data.semanticCommits;
+    data.semanticCommits = 'enabled';
+    data.configFileName = 'renovate.json';
+
+    expect(localRepoCache.isModified()).toBeFalse();
+
+    await localRepoCache.save();
+
+    expect(fs.outputCacheFile).not.toHaveBeenCalledWith();
+  });
```
Comment (zharinov) [inline lib/util/cache/repository/impl/local.spec.ts:342]: Fails with current implementation

**Review by Gabriel-Ladzaretti — COMMENTED** [review]
Ill just note that the `prCahce` [issue](https://github.com/renovatebot/renovate/issues/17603) is unrelated to the sorting problem of objects, although this sorting issue was also observed, so this is very welcome.

~By looking at the source code for `sortObject` it seems that it does sort arrays as well.~

~https://github.com/bevry/sortobject/blob/f8202dc8735030fc82b6840e8cca80502aa3d694/source/index.ts#L53-L58~

Otherwise, code LGTM.

EDIT: This indeed doesnt sort Arrays if no comparator function is provided.

File: lib/util/cache/repository/impl/base.ts [hunk 3]
```diff
@@ -105,8 +106,14 @@ export abstract class RepoCacheBase implements RepoCache {
...
 
+  private stringify(): string {
+    const sortedData = sortObject(this.data);
+    const stringifiedData = JSON.stringify(sortedData);
+    return stringifiedData;
+  }
+
   async save(): Promise<void> {
-    const jsonStr = JSON.stringify(this.data);
+    const jsonStr = this.stringify();
```
Comment (viceice) [inline lib/util/cache/repository/impl/base.ts:116]: i think we should simply use the new fast stable stringify dependency

Comment (zharinov) [reply lib/util/cache/repository/impl/base.ts:116]: Please point me the exact dependency name, I'll add it

Comment (viceice) [reply lib/util/cache/repository/impl/base.ts:116]: https://github.com/renovatebot/renovate/blob/b23b886b6e4c4ac1bcee6ae03a26a60933e28d4f/package.json#L209 🤗

Comment (zharinov) [reply lib/util/cache/repository/impl/base.ts:116]: 🤣🤣🤣

**Review by viceice — APPROVED** [review]

**Pull request merged**